Source/scrollrt.cpp

```diff
@@ -201,14 +201,14 @@ static bool ShouldShowCursor()
  */
 static void DrawCursor(const CelOutputBuffer &out)
 {
-	if (pcurs <= CURSOR_NONE || cursW == 0 || cursH == 0 || !ShouldShowCursor()) {
+	if (pcurs <= CURSOR_NONE || cursSize.width == 0 || cursSize.height == 0 || !ShouldShowCursor()) {

```

Comment:
I actually wonder why that would be even allowed in the first place. We should not allow an invalid rectangle to be created in the first place, so there would be no need for checking.

Something like `Point.Origin` makes more sense to me, but I don't see how an "empty" rectangle is a good thing.

Reply:
SDL provides it to allow for code such as `Rect.Intersection(Rect)` (an example use case is determining the common area between two buffers). Should `Rectangle {0, 0, 30, 30}.Intersection(Rectangle{100, 100, 10, 20})` throw a runtime error or is it nicer to return an empty rect and rely on callers to handle it appropriately?

To use an example from devilution, setting the cursor to CURSOR_NONE calls GetInvSize in a way that makes it create and return a Size with width/height 0. That's a valid use case in the current code, completely avoiding creating a Size object with those properties would require significant refactoring just to handle this one situation.

This also highlights redundant checks in this line. `pcurs <= CURSOR_NONE` is a guard against attempting to draw an invalid cursor. This could be combined with the next condition if we accept that `cursize.isEmpty()` means the cursor should not be drawn. I think it would not be appropriate to combine this with the ShouldShowCursor() check however as that function is more concerned with game state actions that need the cursor hidden temporarily.

Reply:
Well, such intersection methods should probably return `optional<Rectangle>` then, using none to represent "there is no intersection".